File: app/src/main/java/com/eslimaf/mobilechallenge/infrastructure/waveapi/WaveApiService.java
package com.eslimaf.mobilechallenge.infrastructure.waveapi;


import com.eslimaf.mobilechallenge.model.Product;

import java.io.IOException;
import java.util.List;

import okhttp3.Interceptor;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.Response;
import retrofit2.Call;
import retrofit2.Retrofit;
import retrofit2.converter.gson.GsonConverterFactory;
import retrofit2.http.Path;

public class WaveApiService implements WaveApi {
    private static WaveApiService sInstance;
    private WaveApi mWaveApi;

    private WaveApiService() {

        OkHttpClient okHttpClient = new OkHttpClient.Builder()
                .addInterceptor(new Interceptor() {
                    @Override
                    public Response intercept(Chain chain) throws IOException {
                        Request request = chain.request();
                        request = request.newBuilder()
                                .addHeader("Authorization","Bearer 6W9hcvwRvyyZgPu9Odq7ko8DSY8Nfm")
                                .build();
                        return chain.proceed(request);
                    }
                }).build();

        Retrofit retrofit = new Retrofit.Builder()
                .client(okHttpClient)
                .baseUrl(WAVEAPI_BASE_URL)
                .addConverterFactory(GsonConverterFactory.create())
                .build();
        mWaveApi = retrofit.create(WaveApi.class);
    }

    public static synchronized WaveApiService getInstance() {
        if (null == sInstance) {
            sInstance = new WaveApiService();
        }
        return sInstance;
    }

    @Override
    public Call<List<Product>> getProductsByBusinessId(@Path("businessId") String businessId) {
        return mWaveApi.getProductsByBusinessId(businessId);
    }
}

File: app/src/main/java/challenge/waveproductlist/data/WaveApiService.kt
package challenge.waveproductlist.data

import challenge.waveproductlist.BuildConfig
import com.google.gson.FieldNamingPolicy
import com.google.gson.GsonBuilder
import io.reactivex.Single
import okhttp3.Interceptor
import retrofit2.Retrofit
import retrofit2.adapter.rxjava2.RxJava2CallAdapterFactory
import retrofit2.converter.gson.GsonConverterFactory
import retrofit2.http.GET
import retrofit2.http.Query
import okhttp3.OkHttpClient
import okhttp3.Response
import retrofit2.http.Path
import java.io.IOException

interface WaveApiService {

    @GET("businesses/{business_id}/products/")
    fun getBusinessProducts(
        @Path(value = "business_id", encoded = true) businessId: String,
        @Query("page", encoded = true) page: Int,
        @Query("page_size", encoded = true) pageSize: Int) : Single<List<Product>>

    companion object {
        fun getService(): WaveApiService {
            var client = OkHttpClient.Builder().addInterceptor(object : Interceptor {
                @Throws(IOException::class)
                override fun intercept(chain: Interceptor.Chain): Response {
                    val newRequest = chain.request().newBuilder()
                        .addHeader("Authorization", "Bearer " + BuildConfig.apiKey)
                        .build()
                    return chain.proceed(newRequest)
                }
            }).build()

            val gsonBuilder = GsonBuilder()
            gsonBuilder.serializeNulls()
            gsonBuilder.setFieldNamingPolicy(FieldNamingPolicy.LOWER_CASE_WITH_UNDERSCORES)

            val retrofit = Retrofit.Builder()
                .client(client)
                .baseUrl("https://api.waveapps.com/")
                .addCallAdapterFactory(RxJava2CallAdapterFactory.create())
                .addConverterFactory(GsonConverterFactory.create(gsonBuilder.create()))
                .build()
            return retrofit.create(WaveApiService::class.java)
        }
    }
}